open-apparel-registry/open-apparel-registry
# Add facility matching research tools and documentation

## Overview

Adds a `research` directory to hold documents and experimental scripts. The `gazetteer` example represents our intended production address matching workflow, comparing a new, "messy" list of facilities to a canonical list. Specifically, the `oar_gazeteer_example_pass_2.py` script shows running a pre-trained model with a previously unknown set of input data being matched to a canonical list of facilities.

Connects #146
Connects #220

## Notes

These experimental scripts are a proof of concept, not shipping code. They do not include any virtualization or containerization because their scope and dependencies are limited.

## Testing Instructions

* If desired, create and activate a virtualenv.
* `cd research/dedupe`
* Run `bootstrap`. Verify that the script completes without error.
* Run `python dedupe/merge_lists.py data/supplier-lists dedupe/facility_list_items.csv`. Verify that `dedupe/facility_list_items.csv` is created
   * Verify that there are 19214 lines with `wc -l dedupe/facility_list_items.csv`
   * Verify that there is a `id` column.
   * Verify that the country column contains country codes
* `cd dedupe`
* Run `oar_dedupe_example.py`. Walk through the interactive labeling until you get just over 20 positives and just over 20 negatives. Use `f` to finish.
* Verify that `facilities_output.csv` is created. Open it in a spreadsheet program that allows sorting by column and sort by "Cluster ID." Verify that lines with the same Cluster ID and a relatively high `confidence_score` are similar.
* `cd ..` into the `research/dedupe` directory
* Run `python dedupe/merge_lists.py data/ gazetteer/rubota_facilities.csv`
* `cd gazetteer`
* Create "messy" files for the gazetteer example
   * `head -5001 rubota_facilities.csv > gazetteer_messy.csv`
   * `head -1 rubota_facilities.csv > gazetteer_messy_pass_2.csv`
   * `sed -n '7001,11999p;12000q' rubota_facilities.csv >> gazetteer_messy_pass_2.csv`
* `cd ..` to return to the `research/dedupe` directory
* Launch a shell with `docker-compose run --rm research`
* `cd gazetteer`
* Run `python oar_gazeteer_example.py`. Walk through the interactive labeling until you get just over 20 positives and just over 20 negatives. Use `f` to finish.
* Verify that `gazetteer_output.csv` is created. Open it in a spreadsheet program that allows sorting by column and sort by "Link ID." Verify that lines with the same `Link ID` and a relatively high `Link Score` are similar.
* Run `python oar_gazeteer_example_pass_2.py` and verify that it completes without error.
* Verify that `gazetteer_pass_2_output.csv` is created. Open it in a spreadsheet program that allows sorting by column and sort by "Link ID." Verify that lines with the same `Link ID` and a relatively high `Link Score` are similar.
